contracts/cw-native-staked-balance-voting/src/contract.rs

```diff
@@ -0,0 +1,311 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    coins, to_binary, BankMsg, Binary, CosmosMsg, Deps, DepsMut, Env, MessageInfo, Response,
+    StdResult, Uint128,
+};
+use cw2::set_contract_version;
+use cw_controllers::ClaimsResponse;
+use cw_core_interface::voting::{TotalPowerAtHeightResponse, VotingPowerAtHeightResponse};
+use cw_utils::{must_pay, Duration};
+
+use crate::error::ContractError;
+use crate::msg::{ExecuteMsg, InstantiateMsg, MigrateMsg, QueryMsg};
+use crate::state::{Config, CLAIMS, CONFIG, DAO, MAX_CLAIMS, STAKED_BALANCES, STAKED_TOTAL};
+
+const CONTRACT_NAME: &str = "crates.io:cw-native-staked-balance-voting";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+fn validate_duration(duration: Option<Duration>) -> Result<(), ContractError> {
+    if let Some(unstaking_duration) = duration {
+        match unstaking_duration {
+            Duration::Height(height) => {
+                if height == 0 {
+                    return Err(ContractError::InvalidUnstakingDuration {});
+                }
+            }
+            Duration::Time(time) => {
+                if time == 0 {
+                    return Err(ContractError::InvalidUnstakingDuration {});
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let owner = match msg.owner {
+        Some(owner) => Some(deps.api.addr_validate(&owner)?),
+        None => None,
+    };
+
+    let manager = match msg.manager {
+        Some(manager) => Some(deps.api.addr_validate(&manager)?),
+        None => None,
+    };
+
+    validate_duration(msg.unstaking_duration)?;
```

Comment:
Crazy to me that `Duration` is trivially copyable like this. Gets me every time.